optuna_dashboard/ts/components/TrialList.tsx

```diff
@@ -319,20 +320,230 @@ export const TrialListDetail: FC<{
+              left: "50%",
+              transform: "translate(-50%, -50%)",
+              bgcolor: "background.paper",
+              borderRadius: "15px",
+            }}
+          >
+            <ThreejsArtifactViewer
+              src={`/artifacts/${trial.study_id}/${trial.trial_id}/${artifact.artifact_id}`}
+              width={`${innerWidth * 0.8}px`}
+              height={`${innerHeight * 0.8}px`}
+              hasGizmo={true}
+              filetype={artifact.filename.split(".").pop()}
+            />
+          </Box>
+        </Modal>
+      </>
+    )
+  }
+  return null
+}
+
+const TrialArtifact: FC<{
+  trial: Trial
+  artifact: Artifact
+  width: string
+  height: string
+}> = ({ trial, artifact, width, height }) => {
   const [openDeleteArtifactDialog, renderDeleteArtifactDialog] =
     useDeleteArtifactDialog()
+  const theme = useTheme()
+  const is3dModel =
+    artifact.filename.endsWith(".stl") || artifact.filename.endsWith(".3dm")
+  const canDelete = trial.state === "Running" || trial.state === "Waiting"
+  let actionsCount = 1
+  if (canDelete) actionsCount += 1
+  if (is3dModel) actionsCount += 1
+  const actionsWidth = theme.spacing(actionsCount * 4)
+
+  return (
+    <Card
+      key={artifact.artifact_id}
+      sx={{
+        marginBottom: theme.spacing(2),
+        width: width,
+        minHeight: height,
+        margin: theme.spacing(0, 1, 1, 0),
+      }}
+    >
+      <TrialArtifactContent
+        trial={trial}
+        artifact={artifact}
+        width={width}
+        height={height}
+      />
+      <CardContent
+        sx={{
+          display: "flex",
+          flexDirection: "row",
+          padding: `${theme.spacing(1)} !important`,
+        }}
+      >
+        <Typography
+          sx={{
+            p: theme.spacing(0.5, 0),
+            flexGrow: 1,
+            wordWrap: "break-word",
+            maxWidth: `calc(100% - ${actionsWidth})`,
+          }}
+        >
+          {artifact.filename}
+        </Typography>
+        {is3dModel ? (
+          <TrialArtifactActions
+            trial={trial}
+            artifact={artifact}
+            sx={{ margin: "auto 0" }}
+          />
+        ) : null}
+        {canDelete ? (
+          <IconButton
+            aria-label="delete artifact"
+            size="small"
+            color="inherit"
+            sx={{ margin: "auto 0" }}
+            onClick={() => {
+              openDeleteArtifactDialog(trial.study_id, trial.trial_id, artifact)
+            }}
+          >
+            <DeleteIcon />
+          </IconButton>
+        ) : null}
+        <IconButton
+          aria-label="download artifact"
+          size="small"
+          color="inherit"
+          download={artifact.filename}
+          sx={{ margin: "auto 0" }}
+          href={`/artifacts/${trial.study_id}/${trial.trial_id}/${artifact.artifact_id}`}
+        >
+          <DownloadIcon />
+        </IconButton>
+      </CardContent>
+      {renderDeleteArtifactDialog()}
```

Comment:
We don't need to render multiple dialogs actually. Could you update the code to render it only once?